File: backend/app/routes/reviews.py
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import Review
from app.schemas import ReviewCreate, ReviewResponse, ReviewListResponse, Issue
from app.services.ai_service import review_code, LANGUAGE_OPTIONS

router = APIRouter(prefix="/reviews", tags=["reviews"])


@router.get("/languages")
def get_languages():
    return LANGUAGE_OPTIONS


@router.post("", response_model=ReviewResponse, status_code=201)
def create_review(body: ReviewCreate, db: Session = Depends(get_db)):
    if not body.code.strip():
        raise HTTPException(400, "Code cannot be empty")

    issues, summary, score = review_code(body.code, body.language)

    review = Review(
        code=body.code,
        language=body.language,
        title=body.title or f"Review of {body.language} code",
        overall_score=score,
        issues=[i.model_dump() for i in issues],
        summary=summary,
    )
    db.add(review)
    db.commit()
    db.refresh(review)

    return _format_review(review)


@router.get("", response_model=list[ReviewListResponse])
def list_reviews(skip: int = 0, limit: int = 20, db: Session = Depends(get_db)):
    reviews = (
        db.query(Review)
        .order_by(Review.created_at.desc())
        .offset(skip)
        .limit(limit)
        .all()
    )
    return [
        ReviewListResponse(
            id=r.id,
            title=r.title,
            language=r.language,
            overall_score=r.overall_score,
            issue_count=len(r.issues or []),
            created_at=r.created_at,
        )
        for r in reviews
    ]


@router.get("/{review_id}", response_model=ReviewResponse)
def get_review(review_id: str, db: Session = Depends(get_db)):
    review = db.query(Review).filter(Review.id == review_id).first()
    if not review:
        raise HTTPException(404, "Review not found")
    return _format_review(review)


def _format_review(r: Review) -> ReviewResponse:
    return ReviewResponse(
        id=r.id,
        code=r.code,
        language=r.language,
        title=r.title,
        overall_score=r.overall_score,
        issues=[Issue(**i) for i in (r.issues or [])],
        summary=r.summary,
        created_at=r.created_at,
    )

File: backend/src/routes/reviews.ts
import { Router, Request, Response } from "express";
import { Review } from "../models/index.js";
import { reviewCode, LANGUAGE_OPTIONS } from "../services/aiService.js";

const router = Router();

router.get("/languages", (_req: Request, res: Response) => {
  res.json(LANGUAGE_OPTIONS);
});

router.post("/", async (req: Request, res: Response) => {
  try {
    const { code, language, title } = req.body;
    if (!code || !code.trim()) {
      res.status(400).json({ detail: "Code cannot be empty" });
      return;
    }

    const { issues, summary, score } = await reviewCode(code, language || "python");

    const review = new Review({
      code,
      language: language || "python",
      title: title || `Review of ${language || "python"} code`,
      overallScore: score,
      issues,
      summary,
    });
    await review.save();

    res.status(201).json({
      id: review._id,
      code: review.code,
      language: review.language,
      title: review.title,
      overall_score: review.overallScore,
      issues: review.issues,
      summary: review.summary,
      created_at: review.createdAt,
    });
  } catch (err: any) {
    res.status(500).json({ detail: err.message || "Internal server error" });
  }
});

router.get("/", async (req: Request, res: Response) => {
  try {
    const skip = Number(req.query.skip) || 0;
    const limit = Number(req.query.limit) || 20;
    const reviews = await Review.find()
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(limit)
      .lean();

    res.json(
      reviews.map((r) => ({
        id: r._id,
        title: r.title,
        language: r.language,
        overall_score: r.overallScore,
        issue_count: (r.issues || []).length,
        created_at: r.createdAt,
      }))
    );
  } catch (err: any) {
    res.status(500).json({ detail: err.message });
  }
});

router.get("/:id", async (req: Request, res: Response) => {
  try {
    const review = await Review.findById(req.params.id).lean();
    if (!review) {
      res.status(404).json({ detail: "Review not found" });
      return;
    }
    res.json({
      id: review._id,
      code: review.code,
      language: review.language,
      title: review.title,
      overall_score: review.overallScore,
      issues: review.issues,
      summary: review.summary,
      created_at: review.createdAt,
    });
  } catch (err: any) {
    res.status(500).json({ detail: err.message });
  }
});

export default router;
